src/hooks/users/useWorldUsers.tsx

```diff
@@ -0,0 +1,124 @@
+
+import { User } from "types/User";
+
+import { WithId } from "utils/id";
+import { worldUsersWithoutLocationSelector } from "utils/selectors";
+import { isDefined } from "utils/types";
+
+import { isLoaded, useFirestoreConnect } from "hooks/useFirestoreConnect";
+import { useSelector } from "hooks/useSelector";
+import { useSovereignVenueId } from "hooks/useSovereignVenueId";
+
+export interface WorldUsersContextState {
+  isSovereignVenueIdLoading: boolean;
+  sovereignVenueId?: string;
+  sovereignVenueIdError?: string;
+
+  shouldConnect: boolean;
+}
+
+const WorldUsersContext = createContext<WorldUsersContextState | undefined>(
+  undefined
+);
+
+export interface WorldUsersProviderProps {
+  venueId?: string;
+}
+
+export const WorldUsersProvider: React.FC<WorldUsersProviderProps> = ({
+  venueId,
+  children,
+}) => {
+  const {
+    sovereignVenueId,
+    isSovereignVenueIdLoading,
+    errorMsg: sovereignVenueIdError,
+  } = useSovereignVenueId({
+    venueId,
+  });
+
+  const shouldConnect =
+    !isSovereignVenueIdLoading &&
+    isDefined(sovereignVenueId) &&
+    isDefined(venueId);
+
+  // TODO: refactor this to not use useFirestoreConnect
+  useFirestoreConnect(() => {
+    if (!shouldConnect) return [];
+
+    const relatedLocationIds = [venueId];
+
+    if (sovereignVenueId) {
+      relatedLocationIds.push(sovereignVenueId);
+    }
+
+    return [
+      {
+        collection: "users",
+        where: ["enteredVenueIds", "array-contains-any", relatedLocationIds],
+        storeAs: "worldUsers",
+      },
+    ];
+  });
+
+  const worldUsersState: WorldUsersContextState = useMemo(
+    () => ({
+      venueId,
+      isSovereignVenueIdLoading,
+      sovereignVenueId,
+      sovereignVenueIdError,
+      shouldConnect,
+    }),
+    [
+      venueId,
+      isSovereignVenueIdLoading,
+      sovereignVenueId,
+      sovereignVenueIdError,
+      shouldConnect,
+    ]
+  );
+
+  return (
+    <WorldUsersContext.Provider value={worldUsersState}>
+      {children}
+    </WorldUsersContext.Provider>
+  );
+};
+
+export const useWorldUsersContext = (): WorldUsersContextState => {
+  const worldUsersState = useContext(WorldUsersContext);
+
+  if (!worldUsersState) {
+    throw new Error(
+      "<WorldUsersProvider/> not found. Did you forget to include it in your component hierarchy?"
+    );
+  }
+
+  return worldUsersState;
+};
+
+export interface WorldUsersData {
+  worldUsers: readonly WithId<User>[];
+  isWorldUsersLoaded: boolean;
+}
+
+export const useWorldUsers = (): WorldUsersData => {
+  // We mostly use this here to ensure that the WorldUsersProvider has definitely been connected
+  useWorldUsersContext();
+
+  // @debt we use lodash's isEqual here to do a deep compare to prevent re-renders, but ideally we wouldn't need to
+  const selectedWorldUsers = useSelector(
+    worldUsersWithoutLocationSelector,
+    isEqual
+  );
+
+  return useMemo(
+    () => ({
+      worldUsers: selectedWorldUsers ?? [],
```

Comment:
If we keep the `useMemo` here then it's not really needed to extract it into another variable. If we get rid of the `useMemo` then we would need to follow the same pattern to avoid creating a new array every time the hook is called.

I don't think there is any benefit to the `useMemo` here since it would only be of use if we were passing on the entire return object without de-structuring, which we practically never do. `selectedWorldUsers` will already be memo'd via `useSelector`.

So yeah, seems like a decent idea.

Reply:
fixed in c957697a068fcfd596de060d38d9db000e5b8ec3